content/blog/2025-03-27-valkey-8-1-0-rc1.md

```diff
@@ -0,0 +1,257 @@
+    |    |       |
+    |    |      One byte of hash for each entry position in the bucket.
+    |    |
+    |   Presence bits. One bit for each entry position, indicating if an
+    |   entry present or not.
+    |
+   Chained? One bit. If set, the last entry is a child bucket pointer.
+```
+
+The secondary hashes are used for quickly eliminating hash collisions when looking up a key. In this way, we can avoid comparing the key for a mismatching key, except once in 256.
+
+This new design reduces by one the number of memory accesses required to lookup an entry when there are no collisions, but it reduces a lot more memory accesses to lookup an entry when there are key mapping collisions because each bucket can store 7 entries.
+
+It also reduces the memory footprint by a lot when there are collisions. With 3 keys mapping to the same bucket, the new design is already using less 8 bytes than the old design, which requires 72 bytes of memory (3 `dictEntry` objects). In the new design, the 64 byte bucket can store up to 7 entries, while the old design would require 168 bytes.
+
+
+### Other Performance Improvements
+
+Several performance improvements were also added to the 8.1 release, here's a list of a few of them:
+
+#### I/O threads improvements
+
+Following up the I/O threads improvements added in 8.0, more operations have been offloaded to the I/O thread pool in the 8.1 release, improving the throughput and latency of some operations.
+
+In the new 8.1 release, TLS connections are now able to efficiently use I/O threads for the several phases of the protocol. The TLS negotiation, which takes non-negliglibe amount of time to setup new connections, can now be offloaded to the I/O thread pool. This change has improved the rate of accepting new connections by around 300%.
+
+Other sources of overhead in the TLS connection handling were identified, namely in the calls to `SSL_pending()` and `ERR_clear_error()` functions, which were being called in the main event thread. By offloading these functions to the I/O threads pool, a throughput improvement was achieved in some operations. For instance, it was observed a 10% improvement in `SET` operations throughput, and a 22% improvement in `GET` operations throughput.
+
+Replication traffic efficiency was also improved in 8.1 by offloading the reading of replication stream on the replicas to the I/O thread pool which means they can serve more read traffic. On the primaries, replication stream writes are now offloaded to the I/O thread pool.
+
+#### Sorted set and hyperloglog optimisations 
+
+Valkey 8.1 continues to add optimisations which will greatly benefit users. Leaderboards are a popular use case for Valkey using its sorted set data type. Retrieving the specific rank of an entry on the leaderboard is done through the ZRANK operation. 
+
+By analyzing the performance of ZRANK, a new algorithm was developed to optimise it’s performance by up to 45% depending on the size of the sorted set. 
```

Comment:
```suggestion
ZRANK command, which serves a popular usecase in operating Leaderboards, was optimised to perform up to 45% faster, depending on the serted set size.
```

calling it a new "algorithm" makes me feel uncomfortable.

Reply:
It is a new algorithm though. It's some kind of reverse lookup. I remember we wanted to coin a term for this in the PR. I think it deserves a cool term. Let's call it *Ran's algortigm*?